src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/Panel.tsx

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+import React, { useEffect, useState } from "react";
+import classnames from "classnames";
+import PanelEditor from "./PanelEditor";
+import BreakpointNavigation from "devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointNavigation";
+import Widget from "./Widget";
+import { useDispatch, useSelector } from "react-redux";
+import { inBreakpointPanel } from "devtools/client/debugger/src/utils/editor";
+import PanelSummary from "./PanelSummary";
+import FirstEditNag from "./FirstEditNag";
+import hooks from "ui/hooks";
+import { Nag } from "ui/hooks/users";
+import { prefs } from "ui/utils/prefs";
+import { Breakpoint, getHitCountsForSelectedSource } from "devtools/client/debugger/src/selectors";
+import { getAnalysisPointsForBreakpoint } from "ui/reducers/app";
+import { UIState } from "ui/state";
+import { clearHoveredItem, setHoveredItem } from "ui/actions/timeline";
+import { Editor } from "codemirror";
+import { AnalysisError, AnalysisPayload } from "ui/state/app";
+
+export type Input = "condition" | "logValue";
+type CodeMirror = any;
+
+function getPanelWidth(editor: Editor) {
+  // The indent value is an adjustment for the distance from the gutter's left edge
+  // to the panel's left edge, which is approximately ~60.
+  const panelIndent = 60;
+  return editor.getScrollInfo().clientWidth - panelIndent;
+}
+
+export enum PanelErrors {
+  TooManyPoints = "too-many-points",
+  MaxHits = "max-hits",
+  Default = "default",
+}
+
+const getPanelError = (hits: number | null, analysisPoints: AnalysisPayload | null) => {
+  if (analysisPoints?.error === AnalysisError.TooManyPoints) {
+    return PanelErrors.TooManyPoints;
+  } else if (analysisPoints?.error === AnalysisError.Default) {
+    return PanelErrors.Default;
+  } else if (hits && hits > prefs.maxHitsDisplayed) {
+    return PanelErrors.MaxHits;
+  }
+
+  return null;
+};
+
+const useGetHitsForPanel = (breakpoint: Breakpoint) => {
+  const analysisPoints = useSelector((state: UIState) =>
+    getAnalysisPointsForBreakpoint(state, breakpoint)
+  );
+  const hitCounts = useSelector(getHitCountsForSelectedSource);
+  const hitCountFromSources = hitCounts?.find(
+    hitCount => hitCount.location.line === breakpoint.location.line
+  );
```

Comment:
How often do we think this hook will run and with how many `hitCounts` matches? (Wondering if this should be memoized.)

Reply:
it will only run until the first match, but there could be hundreds or thousands of `hitCounts` entries to iterate over).

Reply:
Definitely seems worth. memoizing then rather than doing every render.